### ISSUE TITLE
Request: Add option to not open keyboard on TextField focus

### ISSUE THREAD

**Issue by ghost** [flutter/flutter#16863]
I have often build Android and iOS app's with a custom keyboard that is not really a keyboard but simply a view/widget that is part of the layout, and then handling all input to a given TextField manually in code. To make this work, I need a way to prevent the OS keyboard from showing up when the TextField receives focus. Could a option to disable keyboard for a TextField be added to Flutter?

**Comment by harrisonturton** [comment]
+1 for this.

Example use-case: Facebook Messenger's emoji keyboard. It's not the stock keyboard, but a custom one.

**Comment by zoechi** [comment]
https://docs.flutter.io/flutter/material/TextField/autofocus.html does not do what you want?

**Comment by SemonCat** [comment]
+1

**Comment by kainecy** [comment]
+1

**Comment by udkl** [comment]
+1

**Comment by balthild** [comment]
Please press 👍 on existing posts instald of commenting "+1"s to avoid annoying subscribers.

**Comment by mortonharley** [comment]
Did anyone manage to find a way to disable the keyboard comletely? For my usecase, I am currently hiding it again, after it shows, which looks pretty unclean. Would love to deactivate it on certain inputs before it even shows

**Comment by diegolaballos** [comment]
I found an option to avoid the system keyboard. Basically using this focus node for the textField:
`class AlwaysDisabledFocusNode extends FocusNode {
  @override
  bool get hasFocus => false;
}`
However maybe this is not the ideal solution..

**Comment by josvos** [comment]
+1

In fact, I would like to add another suggestion: would it be an option to add a method to `TextEditingController` to manually "inject" input as if this input was typed on a keyboard, so that this string is inserted at the cursor position?

**Comment by Jardo-51** [comment]
As a workaround, I created my own component which extends EditableText. It uses some hacks to hide the keyboard after it's been shown and also to prevent it showing on autofocus. It's not nice but it does the job. I only put attributes which I needed into the constructor.

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class NoKeyboardEditableText extends EditableText {

  NoKeyboardEditableText({
    @required TextEditingController controller,
    @required TextStyle style,
    @required Color cursorColor,
    bool autofocus = false,
    Color selectionColor
  }):super(
    controller: controller,
    focusNode: NoKeyboardEditableTextFocusNode(),
    style: style,
    cursorColor: cursorColor,
    autofocus: autofocus,
    selectionColor: selectionColor,
    backgroundCursorColor: Colors.black
  );

  @override
  EditableTextState createState() {
    return NoKeyboardEditableTextState();
  }

}

class NoKeyboardEditableTextState extends EditableTextState {
  @override
  void requestKeyboard() {
    super.requestKeyboard();
    //hide keyboard
    SystemChannels.textInput.invokeMethod('TextInput.hide');
  }
}

class NoKeyboardEditableTextFocusNode extends FocusNode {
  @override
  bool consumeKeyboardToken() {
    // prevents keyboard from showing on first focus
    return false;
  }
}
```

**Comment by LampeMW** [comment]
@balthild How would I implement this component into my textfields?

**Comment by Jardo-51** [comment]
> @Jardo-51 @balthild How would I implement this custom component into my textfields?

Just replace `TextField` with `NoKeyboardEditableText` in your code.

**Comment by shubhamhackz** [comment]
Another Option is to simply wrap up the `TextField` inside a `IgnorePointer`, which will basically ignore all touch events to your `TextField` including `Taps`. This solution works for me although this is not the ideal one.

**Comment by josvos** [comment]
This answer is completely unrelated to this issue...

**Comment by PretzelJones** [comment]
After some testing I realized it only solves the problem on the iOS and Android emulators but not hardware.

**Comment by LampeMW** [comment]
@PretzelJones You're trying to solve a different issue than what this is about.  Your solution is about not resizing widgets when the keyboard opens.  This open is is regarding not having the keyboard open when a textfield is focused.

**Comment by PretzelJones** [comment]
> @PretzelJones You're trying to solve a different issue than what this is about. Your solution is about not resizing widgets when the keyboard opens. This open is is regarding not having the keyboard open when a textfield is focused.

I came across the fix I gave when searching for a solution to the keyboard problem. I tried it and it worked on the emulator. 

Anyway, worth a shot. No need to beat it into the ground.

**Comment by rishabh-chimple** [comment]
@Jardo-51  I  used your `NoKeyboardEditableText `  class to select the text and its working fine, but how can I get the text that is being selected ? I am passing the text in `TextEditingController`... also I added 
`  bool get hasFocus => false;` in `NoKeyboardEditableTextFocusNode ` to avoid the underline on selecting the text.

```
class NoKeyboardEditableTextFocusNode extends FocusNode {
  @override
  bool get hasFocus => false;
  @override
  bool consumeKeyboardToken() {
    return false;
  }
}
```


Here is code how I am passing the text in `TextEditingController`  

 ```
NoKeyboardEditableText(
                        controller: TextEditingController(
                          text:
                              "Look at these beautiful horses and elephants! Who brought them here",
                        ),
                        cursorColor: Colors.green,
                        selectionColor: Colors.red,
                        style: TextStyle(
                            fontStyle: FontStyle.normal,
                            fontSize: 30.0,
                            color: Colors.black),
                      ),
```

**Comment by SemonCat** [comment]
@Jardo-51 

Can I upload your solution as a dart library to pub?

**Comment by Jardo-51** [comment]
> but how can I get the text that is being selected ?

The same way as with any other `EditableText`: from the `TextEditingController`

> Can I upload your solution as a dart library to pub?

Yeah why not. Just give me credit ;)

**Comment by hugocbpassos** [comment]
@SemonCat Have you done that? Can you provide us a link?

**Comment by marcglasberg** [comment]
Hey everyone. I suggest a good solution to this problem and related ones would be this: https://github.com/flutter/flutter/issues/31985

**Comment by ymoussaba** [comment]
> ```dart
> @override
>   void requestKeyboard() {
>     super.requestKeyboard();
>     //hide keyboard
>     SystemChannels.textInput.invokeMethod('TextInput.hide');
>   }
> ```

I ended up using your solution and updated the requestKeyboard override like that :
`class NoKeyboardEditableTextState extends EditableTextState {
  @override
  void requestKeyboard() {
    //super.requestKeyboard(); // don't call the super method
    //just request focus
    FocusScope.of(context).requestFocus(widget.focusNode);
  }
}`

**Comment by Jardo-51** [comment]
@ymoussaba I think I was doing it the same way at first but I had some issues with it. I don't remember what kind if issues anymore.

**Comment by akarabach** [comment]
> @Jardo-51 I used your `NoKeyboardEditableText ` class to select the text and its working fine, but how can I get the text that is being selected ? I am passing the text in `TextEditingController`... also I added
> ` bool get hasFocus => false;` in `NoKeyboardEditableTextFocusNode ` to avoid the underline on selecting the text.
> 
> ```dart
> class NoKeyboardEditableTextFocusNode extends FocusNode {
>   @override
>   bool get hasFocus => false;
>   @override
>   bool consumeKeyboardToken() {
>     return false;
>   }
> }
> ```
> 
> Here is code how I am passing the text in `TextEditingController`
> 
> ```dart
> NoKeyboardEditableText(
>                        controller: TextEditingController(
>                          text:
>                              "Look at these beautiful horses and elephants! Who brought them here",
>                        ),
>                        cursorColor: Colors.green,
>                        selectionColor: Colors.red,
>                        style: TextStyle(
>                            fontStyle: FontStyle.normal,
>                            fontSize: 30.0,
>                            color: Colors.black),
>                      ),
> ```

How to paint beautiful material input? I see only ugly text...

**Comment by maheshmnj** [comment]
as a workaround why not design a fake textfield and attach tap widget to it heres what I did,I wanted to show a DatePicker on clicking a textField but as soon as I focused the textField the keyboard appears for few milliseconds and then calls the datePicker so I removed the TextField and designed a fake TextField using a Container
![Peek 2019-08-19 18-40](https://user-images.githubusercontent.com/31410839/63268180-2a86fb80-c2b1-11e9-9a74-db627615654a.gif)


```
        GestureDetector(
          onTap: () async {
            selectDate(context);
          },
          child: Container(
            alignment: Alignment.centerLeft,
            padding: EdgeInsets.only(left: 10),
            margin: EdgeInsets.symmetric(horizontal: 20),
            height: 60,
            child: Text(
              dateHint,
              style: TextStyle(fontSize: 18),
            ),
            decoration: BoxDecoration(
                border: Border.all(width: 1.0, color: formColor),
                borderRadius: BorderRadius.circular(10)),
          ),
        ),

  void selectDate(BuildContext context) async {
    final DateTime datePicked = await showDatePicker(
        context: context,
        initialDate: selectedDate,
        firstDate: DateTime(2000),
        lastDate: DateTime(2020));
    if (datePicked != null) {
      setState(() {
        selectedDate = datePicked;
        dateHint = selectedDate.toString().substring(0, 10);
      });
    }
  }
```

**Comment by chunhtai** [comment]
Have you tried [readonly](https://api.flutter.dev/flutter/material/TextField/readOnly.html)
Different from enabled, readonly still gain focus and will block keyboard connection.

**Comment by lekhang2512** [comment]
> As a workaround, I created my own component which extends EditableText. It uses some hacks to hide the keyboard after it's been shown and also to prevent it showing on autofocus. It's not nice but it does the job. I only put attributes which I needed into the constructor.
> 
> ```dart
> import 'package:flutter/material.dart';
> import 'package:flutter/services.dart';
> 
> class NoKeyboardEditableText extends EditableText {
> 
>   NoKeyboardEditableText({
>     @required TextEditingController controller,
>     @required TextStyle style,
>     @required Color cursorColor,
>     bool autofocus = false,
>     Color selectionColor
>   }):super(
>     controller: controller,
>     focusNode: NoKeyboardEditableTextFocusNode(),
>     style: style,
>     cursorColor: cursorColor,
>     autofocus: autofocus,
>     selectionColor: selectionColor,
>     backgroundCursorColor: Colors.black
>   );
> 
>   @override
>   EditableTextState createState() {
>     return NoKeyboardEditableTextState();
>   }
> 
> }
> 
> class NoKeyboardEditableTextState extends EditableTextState {
>   @override
>   void requestKeyboard() {
>     super.requestKeyboard();
>     //hide keyboard
>     SystemChannels.textInput.invokeMethod('TextInput.hide');
>   }
> }
> 
> class NoKeyboardEditableTextFocusNode extends FocusNode {
>   @override
>   bool consumeKeyboardToken() {
>     // prevents keyboard from showing on first focus
>     return false;
>   }
> }
> ```

Bạn có thể cho biết làm cách nào để nhận giá trị từ NoKeyboardEditableText. Tôi đang sử dụng máy quét mã vạch

**Comment by lekhang2512** [comment]
Đây là widget scan qrcode của tôi . Nó làm việc cho tôi
`
import 'package:app_exploited/bloc/navigation/bloc.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

class PdaScanSection extends StatelessWidget {
  final ValueChanged<String> onChanged;
  final NavigationEvent eventBack;

  PdaScanSection({Key key, ValueChanged<String> onChanged, NavigationEvent eventBack})
      : onChanged = onChanged,
        eventBack = eventBack != null ? eventBack : HomeNavigationEvent(),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return MainSection(onChanged: onChanged, eventBack: eventBack);
  }
}

class MainSection extends StatefulWidget {
  final ValueChanged<String> onChanged;
  final NavigationEvent eventBack;

  MainSection({
    Key key,
    @required this.onChanged,
    @required this.eventBack,
  }) : assert(onChanged != null), assert(eventBack != null), super(key: key);

  @override
  State<MainSection> createState() => _MainSectionState(onChanged: onChanged, eventBack: eventBack);
}

class _MainSectionState extends State<MainSection> {
  ValueChanged<String> onChanged;
  NavigationEvent eventBack;
  NavigationBloc navigationBloc;
  TextEditingController scanController = TextEditingController();
  FocusNode myFocusNode;

  _MainSectionState({
    Key key,
    @required this.onChanged,
    @required this.eventBack,
  }) : assert(onChanged != null), assert(eventBack != null);

  @override
  void initState() {
    scanController.addListener(() {
      SystemChannels.textInput.invokeMethod('TextInput.hide');
    });
    super.initState();
    navigationBloc = BlocProvider.of<NavigationBloc>(context);
    myFocusNode = FocusNode();
  }

  @override
  void dispose() {
    myFocusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.blue,
      child: Column(
        children: <Widget>[
          SizedBox(
            height: 25,
          ),
          SizedBox(
            height: 60,
            child: Padding(
              padding: const EdgeInsets.only(right: 20.0),
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: <Widget>[
                      Expanded(
                        flex: 1,
                        child: IconButton(
                          icon: Icon(Icons.arrow_back),
                          color: Colors.white,
                          onPressed: () {
                            navigationBloc.dispatch(eventBack);
                          },
                        ),
                      ),
                      Expanded(
                        flex: 5,
                        child: TextField(
                          focusNode: myFocusNode,
                          controller: scanController,
                          autofocus: true,
                          onChanged: (code) {
                            scanController.text = '';
                            onChanged(code);
                            FocusScope.of(context).requestFocus(myFocusNode);
                          },
                          onTap: () {
                            SystemChannels.textInput.invokeMethod('TextInput.hide');
                          },
                          onEditingComplete: () {
                            SystemChannels.textInput.invokeMethod('TextInput.hide');
                          },
                          style: TextStyle(color: Colors.white),
                          textAlign: TextAlign.center,
                          decoration: InputDecoration(
                            fillColor: Colors.white,
                            focusedBorder: OutlineInputBorder(
                                borderSide: BorderSide(color: Colors.white, width: 2.0)
                            ),
                            hasFloatingPlaceholder: false,
                            labelText: 'Quét mã',
                            labelStyle: TextStyle(color: Colors.white),
                            isDense: true,
                            contentPadding: EdgeInsets.all(10),
                            border: OutlineInputBorder(
                              borderRadius: BorderRadius.all(Radius.circular(5.0)),
                            ),
                            enabledBorder: OutlineInputBorder(
                              borderSide: BorderSide(color: Colors.white, width: 2.0),
                            ),
                          ),
                        )
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ),
        ],
      )
    );
  }
}
`

**Comment by CrazyAvak** [comment]
> Have you tried [readonly](https://api.flutter.dev/flutter/material/TextField/readOnly.html)
> Different from enabled, readonly still gain focus and will block keyboard connection.

thanks this worked perfectly with an ontap to open the calendar

**Comment by n0tc0ding** [comment]
> Have you tried [readonly](https://api.flutter.dev/flutter/material/TextField/readOnly.html)
> Different from enabled, readonly still gain focus and will block keyboard connection.

This worked for me with showDatePicker.

**Comment by mdc555** [comment]
I also need this in order to show a custom keyboard. My problem with a fake TextField and with disabling the normal TextField is that I will need to be able to edit the text on that TextField by dragging the cursor and to input at the actual cursor position.... :(

**Comment by andersonptf-zion** [comment]
> This is my qrcode scan widget. It worked for me
> 
> ```
> import 'package:flutter/material.dart';
> import 'package:flutter/services.dart';
> import 'package:flutter_bloc/flutter_bloc.dart';
> 
> class PdaScanSection extends StatelessWidget {
>   final ValueChanged<String> onChanged;
> 
>   PdaScanSection({Key key, ValueChanged<String> onChanged})
>       : onChanged = onChanged,
>         super(key: key);
> 
>   @override
>   Widget build(BuildContext context) {
>     return MainSection(onChanged: onChanged);
>   }
> }
> 
> class MainSection extends StatefulWidget {
>   final ValueChanged<String> onChanged;
> 
>   MainSection({
>     Key key,
>     @required this.onChanged,
>   }) : assert(onChanged != null), super(key: key);
> 
>   @override
>   State<MainSection> createState() => _MainSectionState(onChanged: onChanged);
> }
> 
> class _MainSectionState extends State<MainSection> {
>   ValueChanged<String> onChanged;
>   TextEditingController scanController = TextEditingController();
>   FocusNode myFocusNode;
> 
>   _MainSectionState({
>     Key key,
>     @required this.onChanged,
>   }) : assert(onChanged != null);
> 
>   @override
>   void initState() {
>     scanController.addListener(() {
>       SystemChannels.textInput.invokeMethod('TextInput.hide');
>     });
>     super.initState();
>     myFocusNode = FocusNode();
>   }
> 
>   @override
>   void dispose() {
>     myFocusNode.dispose();
>     super.dispose();
>   }
> 
>   @override
>   Widget build(BuildContext context) {
>     return Container(
>       color: Colors.blue,
>       child: Column(
>         children: <Widget>[
>           SizedBox(
>             height: 25,
>           ),
>           SizedBox(
>             height: 60,
>             child: Padding(
>               padding: const EdgeInsets.only(right: 20.0),
>               child: Column(
>                 mainAxisAlignment: MainAxisAlignment.center,
>                 children: <Widget>[
>                   Row(
>                     mainAxisAlignment: MainAxisAlignment.spaceBetween,
>                     crossAxisAlignment: CrossAxisAlignment.center,
>                     children: <Widget>[
>                       Expanded(
>                         flex: 1,
>                         child: IconButton(
>                           icon: Icon(Icons.arrow_back),
>                           color: Colors.white,
>                           onPressed: () {
>                             
>                           },
>                         ),
>                       ),
>                       Expanded(
>                         flex: 5,
>                         child: TextField(
>                           focusNode: myFocusNode,
>                           controller: scanController,
>                           autofocus: true,
>                           onChanged: (code) {
>                             scanController.text = '';
>                             onChanged(code);
>                             FocusScope.of(context).requestFocus(myFocusNode);
>                           },
>                           onTap: () {
>                             SystemChannels.textInput.invokeMethod('TextInput.hide');
>                           },
>                           onEditingComplete: () {
>                             SystemChannels.textInput.invokeMethod('TextInput.hide');
>                           },
>                           style: TextStyle(color: Colors.white),
>                           textAlign: TextAlign.center,
>                           decoration: InputDecoration(
>                             fillColor: Colors.white,
>                             focusedBorder: OutlineInputBorder(
>                                 borderSide: BorderSide(color: Colors.white, width: 2.0)
>                             ),
>                             hasFloatingPlaceholder: false,
>                             labelText: 'Quét mã',
>                             labelStyle: TextStyle(color: Colors.white),
>                             isDense: true,
>                             contentPadding: EdgeInsets.all(10),
>                             border: OutlineInputBorder(
>                               borderRadius: BorderRadius.all(Radius.circular(5.0)),
>                             ),
>                             enabledBorder: OutlineInputBorder(
>                               borderSide: BorderSide(color: Colors.white, width: 2.0),
>                             ),
>                           ),
>                         )
>                       ),
>                     ],
>                   ),
>                 ],
>               ),
>             ),
>           ),
>         ],
>       )
>     );
>   }
> }
> ```

@lekhang2512 I'm also using a barcode scanner but your solution hides the keyboard for all the TextFields in the screen. Is there a way to hide the keyboard for a specific field?

**Comment by mikeesouth** [comment]
> I also need this in order to show a custom keyboard. My problem with a fake TextField and with disabling the normal TextField is that I will need to be able to edit the text on that TextField by dragging the cursor and to input at the actual cursor position.... :(

I'm looking for exactly this. I also tried setting `getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN);` in native Java code for Android. This doesn't seem to have any effect in my emulated tablet (Pixel C) or my phone (Huawei Mate 20 Pro). I also tried SOFT_INPUT_STATE_ALWAYS_HIDDEN - does not help.

Not even setting `android:windowSoftInputMode="stateHidden"` (or `stateAlwaysHidden`) in the manifest seems to help. Even if it would work it wouldn't be a great solution for me as I want to have different views, some with the soft input enabled and some with it disabled.

Anyone have ideas?

**Comment by chunhtai** [comment]
With my previous suggestion https://github.com/flutter/flutter/issues/16863#issuecomment-522745415
is there any other use case that is not covered?

**Comment by hugocbpassos** [comment]
@chunhtai Yes. Imagine you have a chat

**Comment by hugocbpassos** [comment]
@chunhtai Yes. Imagine you have a chat-screen, in which you want the cursor to blink in the text-field as soon as the screen opens, just like WhatsApp does. We can do that by focusing on the text-field, but then the keyboard opens. Passing `readonly: true` doesn't help at all, since it prevents the cursor from showing up.

**Comment by chunhtai** [comment]
@hugocbpassos howabout readOnly: true  and [showCursor](https://api.flutter.dev/flutter/material/TextField/showCursor.html): true

**Comment by hugocbpassos** [comment]
@chunhtai Sorry, I missed that one. I can't see any hindrance when using both features.

**Comment by dnfield** [comment]
Closing as resolved.

**Comment by yangfanyu** [comment]
> > This is my qrcode scan widget. It worked for me
> > ```
> > import 'package:flutter/material.dart';
> > import 'package:flutter/services.dart';
> > import 'package:flutter_bloc/flutter_bloc.dart';
> > 
> > class PdaScanSection extends StatelessWidget {
> >   final ValueChanged<String> onChanged;
> > 
> >   PdaScanSection({Key key, ValueChanged<String> onChanged})
> >       : onChanged = onChanged,
> >         super(key: key);
> > 
> >   @override
> >   Widget build(BuildContext context) {
> >     return MainSection(onChanged: onChanged);
> >   }
> > }
> > 
> > class MainSection extends StatefulWidget {
> >   final ValueChanged<String> onChanged;
> > 
> >   MainSection({
> >     Key key,
> >     @required this.onChanged,
> >   }) : assert(onChanged != null), super(key: key);
> > 
> >   @override
> >   State<MainSection> createState() => _MainSectionState(onChanged: onChanged);
> > }
> > 
> > class _MainSectionState extends State<MainSection> {
> >   ValueChanged<String> onChanged;
> >   TextEditingController scanController = TextEditingController();
> >   FocusNode myFocusNode;
> > 
> >   _MainSectionState({
> >     Key key,
> >     @required this.onChanged,
> >   }) : assert(onChanged != null);
> > 
> >   @override
> >   void initState() {
> >     scanController.addListener(() {
> >       SystemChannels.textInput.invokeMethod('TextInput.hide');
> >     });
> >     super.initState();
> >     myFocusNode = FocusNode();
> >   }
> > 
> >   @override
> >   void dispose() {
> >     myFocusNode.dispose();
> >     super.dispose();
> >   }
> > 
> >   @override
> >   Widget build(BuildContext context) {
> >     return Container(
> >       color: Colors.blue,
> >       child: Column(
> >         children: <Widget>[
> >           SizedBox(
> >             height: 25,
> >           ),
> >           SizedBox(
> >             height: 60,
> >             child: Padding(
> >               padding: const EdgeInsets.only(right: 20.0),
> >               child: Column(
> >                 mainAxisAlignment: MainAxisAlignment.center,
> >                 children: <Widget>[
> >                   Row(
> >                     mainAxisAlignment: MainAxisAlignment.spaceBetween,
> >                     crossAxisAlignment: CrossAxisAlignment.center,
> >                     children: <Widget>[
> >                       Expanded(
> >                         flex: 1,
> >                         child: IconButton(
> >                           icon: Icon(Icons.arrow_back),
> >                           color: Colors.white,
> >                           onPressed: () {
> >                             
> >                           },
> >                         ),
> >                       ),
> >                       Expanded(
> >                         flex: 5,
> >                         child: TextField(
> >                           focusNode: myFocusNode,
> >                           controller: scanController,
> >                           autofocus: true,
> >                           onChanged: (code) {
> >                             scanController.text = '';
> >                             onChanged(code);
> >                             FocusScope.of(context).requestFocus(myFocusNode);
> >                           },
> >                           onTap: () {
> >                             SystemChannels.textInput.invokeMethod('TextInput.hide');
> >                           },
> >                           onEditingComplete: () {
> >                             SystemChannels.textInput.invokeMethod('TextInput.hide');
> >                           },
> >                           style: TextStyle(color: Colors.white),
> >                           textAlign: TextAlign.center,
> >                           decoration: InputDecoration(
> >                             fillColor: Colors.white,
> >                             focusedBorder: OutlineInputBorder(
> >                                 borderSide: BorderSide(color: Colors.white, width: 2.0)
> >                             ),
> >                             hasFloatingPlaceholder: false,
> >                             labelText: 'Quét mã',
> >                             labelStyle: TextStyle(color: Colors.white),
> >                             isDense: true,
> >                             contentPadding: EdgeInsets.all(10),
> >                             border: OutlineInputBorder(
> >                               borderRadius: BorderRadius.all(Radius.circular(5.0)),
> >                             ),
> >                             enabledBorder: OutlineInputBorder(
> >                               borderSide: BorderSide(color: Colors.white, width: 2.0),
> >                             ),
> >                           ),
> >                         )
> >                       ),
> >                     ],
> >                   ),
> >                 ],
> >               ),
> >             ),
> >           ),
> >         ],
> >       )
> >     );
> >   }
> > }
> > ```
> 
> @lekhang2512 I'm also using a barcode scanner but your solution hides the keyboard for all the TextFields in the screen. Is there a way to hide the keyboard for a specific field?

`
import 'package:flutter/services.dart';
......
SystemChannels.textInput.invokeMethod('TextInput.hide');
.....

SystemChannels.textInput.invokeMethod('TextInput.show');
`

Perfect for me, thanks

**Comment by andrespd99** [comment]
> This is my qrcode scan widget. It worked for me
> 
> ```
> import 'package:flutter/material.dart';
> import 'package:flutter/services.dart';
> import 'package:flutter_bloc/flutter_bloc.dart';
> 
> class PdaScanSection extends StatelessWidget {
>   final ValueChanged<String> onChanged;
> 
>   PdaScanSection({Key key, ValueChanged<String> onChanged})
>       : onChanged = onChanged,
>         super(key: key);
> 
>   @override
>   Widget build(BuildContext context) {
>     return MainSection(onChanged: onChanged);
>   }
> }
> 
> class MainSection extends StatefulWidget {
>   final ValueChanged<String> onChanged;
> 
>   MainSection({
>     Key key,
>     @required this.onChanged,
>   }) : assert(onChanged != null), super(key: key);
> 
>   @override
>   State<MainSection> createState() => _MainSectionState(onChanged: onChanged);
> }
> 
> class _MainSectionState extends State<MainSection> {
>   ValueChanged<String> onChanged;
>   TextEditingController scanController = TextEditingController();
>   FocusNode myFocusNode;
> 
>   _MainSectionState({
>     Key key,
>     @required this.onChanged,
>   }) : assert(onChanged != null);
> 
>   @override
>   void initState() {
>     scanController.addListener(() {
>       SystemChannels.textInput.invokeMethod('TextInput.hide');
>     });
>     super.initState();
>     myFocusNode = FocusNode();
>   }
> 
>   @override
>   void dispose() {
>     myFocusNode.dispose();
>     super.dispose();
>   }
> 
>   @override
>   Widget build(BuildContext context) {
>     return Container(
>       color: Colors.blue,
>       child: Column(
>         children: <Widget>[
>           SizedBox(
>             height: 25,
>           ),
>           SizedBox(
>             height: 60,
>             child: Padding(
>               padding: const EdgeInsets.only(right: 20.0),
>               child: Column(
>                 mainAxisAlignment: MainAxisAlignment.center,
>                 children: <Widget>[
>                   Row(
>                     mainAxisAlignment: MainAxisAlignment.spaceBetween,
>                     crossAxisAlignment: CrossAxisAlignment.center,
>                     children: <Widget>[
>                       Expanded(
>                         flex: 1,
>                         child: IconButton(
>                           icon: Icon(Icons.arrow_back),
>                           color: Colors.white,
>                           onPressed: () {
>                             
>                           },
>                         ),
>                       ),
>                       Expanded(
>                         flex: 5,
>                         child: TextField(
>                           focusNode: myFocusNode,
>                           controller: scanController,
>                           autofocus: true,
>                           onChanged: (code) {
>                             scanController.text = '';
>                             onChanged(code);
>                             FocusScope.of(context).requestFocus(myFocusNode);
>                           },
>                           onTap: () {
>                             SystemChannels.textInput.invokeMethod('TextInput.hide');
>                           },
>                           onEditingComplete: () {
>                             SystemChannels.textInput.invokeMethod('TextInput.hide');
>                           },
>                           style: TextStyle(color: Colors.white),
>                           textAlign: TextAlign.center,
>                           decoration: InputDecoration(
>                             fillColor: Colors.white,
>                             focusedBorder: OutlineInputBorder(
>                                 borderSide: BorderSide(color: Colors.white, width: 2.0)
>                             ),
>                             hasFloatingPlaceholder: false,
>                             labelText: 'Quét mã',
>                             labelStyle: TextStyle(color: Colors.white),
>                             isDense: true,
>                             contentPadding: EdgeInsets.all(10),
>                             border: OutlineInputBorder(
>                               borderRadius: BorderRadius.all(Radius.circular(5.0)),
>                             ),
>                             enabledBorder: OutlineInputBorder(
>                               borderSide: BorderSide(color: Colors.white, width: 2.0),
>                             ),
>                           ),
>                         )
>                       ),
>                     ],
>                   ),
>                 ],
>               ),
>             ),
>           ),
>         ],
>       )
>     );
>   }
> }
> ```

It worked for me. My use case is exactly the same. I'm working with a barcode scanner. Thanks for your contribution.

**Comment by amreniouinnovent** [comment]
works for me
```
            enableInteractiveSelection: !widget.disableKeyboard,
            onTap: () {
              if(widget.disableKeyboard)
              FocusScope.of(context).requestFocus(new FocusNode());
            },
           // readOnly: widget.disableKeyboard,
            //showCursor: !widget.disableKeyboard,

```

**Comment by fanchou** [comment]
#69146

**Comment by joshpachner** [comment]
For what it is worth this is what i have to help toggle the keyboard between tabs
i start with a var
          var _keyboardHidden = false;
     //then i have a function that will call textinput.show or textinput.hide
     void _toggleKeyboard() {
    print("Trying to hide keybaord!");
    if (_keyboardHidden) {
      SystemChannels.textInput.invokeMethod('TextInput.show');
      _keyboardHidden = false;
    } else {
      SystemChannels.textInput.invokeMethod('TextInput.hide');
      _keyboardHidden =true;
    }
  }

/// and then on my text field i can the function 
 onTap: () => _toggleKeyboard(),


i would post the whole code but its alot, and  that might be more confusing 
basically what i did is use a boolean to know if keyboard has been dismissed and then if its been dismissed on tab it will open it back up and then vice versa. 
hopefully this helps someone!

**Comment by flutterq** [comment]
Step By Step Guide to [How to disable text edit field in flutter ?](https://flutterq.com/wp-admin/post.php?post=243&action=edit)

**Comment by vincent2yui** [comment]
Where's the article in your blog? It goes directly to a WordPress login page and I can't find any info related to the issue.

**Comment by joshpachner** [comment]
What?

**Comment by nicolasknz** [comment]
I tried a lot of stuff, the only thing that solved my problem for good was the lib "input_with_keyboard_control", you can find it at pub.dev.

It has a class InputWithKeyboardControl and you can choose whether you want to hide the keyboard or not, highly recommend it.

**Comment by ycherniavskyi** [comment]
It seems that this issue related to #69146 and when it will merge we will have the necessary functionality to prevent open keyboard within the flutter.

**Comment by NachiketGohil** [comment]
### I have some working suggestion for this,
I am using `TextField` because I want to use it's attributes and also I need to hide keyboard all the time for my custom keyboard.
so what i did is, added just 2 attributes in it:

```
TextField(
    showCursor: true,
    readOnly: true,
);
```
_Hope it's helpful to Someone..._
**PS: I am using Flutter 2.0**

**Comment by Danishkhan007** [comment]
> @Jardo-51 I used your `NoKeyboardEditableText ` class to select the text and its working fine, but how can I get the text that is being selected ? I am passing the text in `TextEditingController`... also I added
> ` bool get hasFocus => false;` in `NoKeyboardEditableTextFocusNode ` to avoid the underline on selecting the text.
> 
> ```dart
> class NoKeyboardEditableTextFocusNode extends FocusNode {
>   @override
>   bool get hasFocus => false;
>   @override
>   bool consumeKeyboardToken() {
>     return false;
>   }
> }
> ```
> 
> Here is code how I am passing the text in `TextEditingController`
> 
> ```dart
> NoKeyboardEditableText(
>                        controller: TextEditingController(
>                          text:
>                              "Look at these beautiful horses and elephants! Who brought them here",
>                        ),
>                        cursorColor: Colors.green,
>                        selectionColor: Colors.red,
>                        style: TextStyle(
>                            fontStyle: FontStyle.normal,
>                            fontSize: 30.0,
>                            color: Colors.black),
>                      ),
> ```

add function as parameter in custom widget and call the function during onChanged method of textfield

**Comment by Danishkhan007** [comment]
> As a workaround, I created my own component which extends EditableText. It uses some hacks to hide the keyboard after it's been shown and also to prevent it showing on autofocus. It's not nice but it does the job. I only put attributes which I needed into the constructor.
> 
> ```dart
> import 'package:flutter/material.dart';
> import 'package:flutter/services.dart';
> 
> class NoKeyboardEditableText extends EditableText {
> 
>   NoKeyboardEditableText({
>     @required TextEditingController controller,
>     @required TextStyle style,
>     @required Color cursorColor,
>     bool autofocus = false,
>     Color selectionColor
>   }):super(
>     controller: controller,
>     focusNode: NoKeyboardEditableTextFocusNode(),
>     style: style,
>     cursorColor: cursorColor,
>     autofocus: autofocus,
>     selectionColor: selectionColor,
>     backgroundCursorColor: Colors.black
>   );
> 
>   @override
>   EditableTextState createState() {
>     return NoKeyboardEditableTextState();
>   }
> 
> }
> 
> class NoKeyboardEditableTextState extends EditableTextState {
>   @override
>   void requestKeyboard() {
>     super.requestKeyboard();
>     //hide keyboard
>     SystemChannels.textInput.invokeMethod('TextInput.hide');
>   }
> }
> 
> class NoKeyboardEditableTextFocusNode extends FocusNode {
>   @override
>   bool consumeKeyboardToken() {
>     // prevents keyboard from showing on first focus
>     return false;
>   }
> }
> ```

Any additional changes for iOS ? it works with Android not with iOS

**Comment by jpnurmi** [comment]
FYI, `TextInputType.none` was introduced in https://github.com/flutter/flutter/pull/83974:
```
TextField(
  keyboardType: TextInputType.none,
  ...
)
```
